feat(api): add support for lua function & description in keymap

shadmansaleh opened the pull request:
### Added support for lua function in keymaps in
--------------------------------------------
- nvim_set_keymap
  Can set lua function as keymap rhs like following:
```lua
vim.api.nvim_{buf_}set_keymap('n', '<leader>lr', '', {callback = vim.lsp.buf.references})
```
  Note: lua function can only be set from lua . If api function being
  called from viml or over rpc this option isn't available.
- nvim_{buf_}get_keymap
  When called from lua, lua function is returned is `callback` key .
  But in other cases callback contains number of the function ref.
- :umap, nvim_del_keymap & nvim_buf_del_keymap clears lua keymaps correctly.
- :map commands for displaing rhs .
   For lua keymaps rhs is displayed as <Lua function ref_no>
   Note: lua keymap cannot be set through viml command / functions.
- mapargs()
  When dict is false it returns string in `<Lua function ref_no>`
  format (same format as :map commands).
  When dict is true it returns ref_no number in `callback` key.
- mapcheck()
  returns string in `<Lua function ref_no>` format (same format as :map commands).

### Added support for keymap description
---------------------------------------
- nvim_{buf_}set_keymap: added `desc` option in opts table .
 ```lua
vim.api.nvim_set_keymap('n', '<leader>w', '<cmd>w<cr>', {desc='Save current file'})
```
- nvim_{buf_}get_keymap: contains `desc` in returned list.
- commands like `:nmap <leader>w` will show description in a new line below rhs.
- `maparg()` return dict contains `desc`.

closes #14976
closes #6123

bfredl commented on src/nvim/api/private/helpers.c:
```diff
@@ -604,6 +605,9 @@ void modify_keymap(Buffer buffer, bool is_unmap, String mode, String lhs, String
     return;
   }
 
+  if (channel_id == LUA_INTERNAL_CALL && opts != NULL && opts->func.type == kObjectTypeLuaRef) {
```
no need to check `channel_id` , only a lua caller can produce a `kObjectTypeLuaRef` (and if there was a way, it wouldn't be a problem either..)

shadmansaleh replied:
Ok I'll remove those.

numToStr commented:
> * nvim_set_keymaps
> * nvim_buf_set_keymaps
>
> * nvim_get_keymaps
> * nvim_buf_get_keymaps

@shadmansaleh I suppose these should be singular i.e `nvim_set_keymap`?

shadmansaleh commented:
> @shadmansaleh I suppose these should be singular i.e nvim_set_keymap?

These are already existing api functions. I'm just extending their functionality .

numToStr commented:
I was talking about the typo in the description.

gpanders commented on src/nvim/api/private/helpers.c:
```diff
@@ -604,6 +605,9 @@ void modify_keymap(Buffer buffer, bool is_unmap, String mode, String lhs, String
     return;
   }
 
+  if (opts != NULL && opts->func.type == kObjectTypeLuaRef) {
```
>And this is a helper function for api not an api function, I'm not sure what you meant by it will always be defined per the API contract.

Yep, you're right! This was the part I missed in my review :)

shadmansaleh commented on src/nvim/getchar.c:
```diff
@@ -3925,15 +3961,29 @@ static char_u *eval_map_expr(char_u *str, int c)
+      }
+      api_free_object(ret);
+    } else {
+      nlua_call_ref(mp->m_luaref, NULL, args, false, &err);
```
Shouldn't we through error if mapped function causes an error ? It's not implemented yet . But I do want to through error here.

gpanders replied:
Yes I think so too.

shadmansaleh commented on src/nvim/getchar.c:
```diff
@@ -3925,15 +3960,29 @@ static char_u *eval_map_expr(char_u *str, int c)
+    if (mp->m_expr) {
+      Object ret = nlua_call_ref(mp->m_luaref, NULL, args, true, &err);
+      if (ret.type == kObjectTypeString) {
+        p = (char_u *)xstrndup(ret.data.string.data, ret.data.string.size);
```
Should we apply `replace_termcodes` on result of function ?

bfredl replied:
Too magic. I'd rather add a shortcut for `vim.api.nvim_replace_termcodes(str, true, true, true)` (like `vim.termcode(str)` or something ) in the next PR.

shadmansaleh commented on src/nvim/api/private/helpers.c:
```diff
@@ -1069,8 +1083,18 @@ ArrayOf(Dictionary) keymap_array(String mode, buf_T *buf)
+        //   Dictionary d = api_dict.data.dictionary;
+        //   for (size_t j = 0; j < d.size; j++) {
+        //     if (strequal("luaref", d.items[j].key.data)) {
+        //       PUT(d, "func", LUAREF_OBJ((LuaRef)d.items[j].value.data.integer));
```
This doesn't work if I try to put anything in the dict (I tried even putting a bool) I get a `string length overflow` error from `luajit` . Any idea why ?

shadmansaleh commented:
Waiting for review . I think I've got everything related to keymaps (If there's somthing left let me know). See https://github.com/neovim/neovim/pull/16594#issue-798921436 for summary on behavioral changes .

shadmansaleh commented:
> The luaref function executes in textlock mode, so many things someone might like to do require using vim.schedule

Regular expr mappings are ran under textlock that's why lua functions are also ran under textlock . I assume there's a reason why exprs are ran under textlock and running without textlock may cause unexpected issues . Though in this case others will know better then me .

> I do think we should add a note to the docs explaining that the luaref function is called with textlock active.

Indeed.

> Even if I use vim.schedule in the above example, I still have to press another key after entering ]g for the mapping to execute

You're right. It seems like something is blocking after function is ran. I need to read more of how keymaps are handled. If any one has a clue on what might be blocking please let me know 😄

> letting the rhs argument be nil when luaref is present

The internal code doesn't care whether rhs is nil or '' as for lua keymaps NULL is stored for rhs . What's blocking is lua api bindings . Those are generated from c function definitions . I'm not completely sure how to make them accept multiple types  🤔

zeertzjq commented:
I think `<expr>` mapping are run under textlock because it isn't expected for `<expr>` to have side-effects. `<expr>` is just used for evaluation, not for actual action. Meanwhile Lua mappings should be more similar to `<Cmd>` mappings.

zeertzjq commented:
I feel for non-`<expr>` LuaRef mappings, there can be an alternative implementation: encode a LuaRef as a keycode and insert it into the typeahead buffer. It can be later handled like `K_COMMAND`.

shadmansaleh commented:
I think we can allow non-expr lua keymaps to execute out of textlock . Also I'll take a look at how K_COMMAND works.

bfredl commented:
Yes, lua mappings which are not `expr=true` need to use some mechanism similar to / based on `K_COMMAND`.

shadmansaleh commented:
@bfredl is `K_COMMAND` implementation isolated in https://github.com/neovim/neovim/pull/4419 ?

zeertzjq commented:
BTW, `K_COMMAND` only appears 15 times in source code.

bfredl commented:
Yes, you find all places where `K_COMMAND` needs to be handled (one per interactive mode) by grepping for it :]

seblj commented:
I have been following this PR, and the one for commands https://github.com/neovim/neovim/pull/16752

@bfredl suggested to have the luaref as part of the `command` in the other PR, instead og having it in the opts table. Personally I like the approach he suggested more than specifying it in the opts table. So I was just wondering if it is possible to do here as well?

zeertzjq commented:
I think this will break backward-compatibility.

lewis6991 commented:
> I think this will break backward-compatibility.

How?

zeertzjq commented:
> > I think this will break backward-compatibility.
> 
> How?

Oh, I didn't look at that thread carefully.

gpanders commented on src/nvim/api/private/helpers.c:
```diff
@@ -605,8 +606,13 @@ void modify_keymap(Buffer buffer, bool is_unmap, String mode, String lhs, String
 
-  if (opts != NULL && opts->luaref.type == kObjectTypeLuaRef) {
-    lua_funcref = api_new_luaref(opts->luaref.data.luaref);
+  if (rhs_obj.type == kObjectTypeString) {
+    rhs = rhs_obj.data.string;
+  } else if (rhs_obj.type == kObjectTypeLuaRef) {
+    lua_funcref = api_new_luaref(rhs_obj.data.luaref);
+  } else {
+    api_set_error(err, kErrorTypeValidation, "Unexpected type of rhs provided");
+    return;
```
I would recommend also adding a `desc` option for when `rhs` is a function (similar to #16752) that is listed when users use `:map {lhs}`.

shadmansaleh replied:
Good idea . I was initially thinking of adding it on a followup . But can be added right here .

shadmansaleh commented:
> Yes, lua mappings which are not expr=true need to use some mechanism similar to / based on K_COMMAND.

@bfredl why though ? In case of K_COMMAND I can see that those maps need to be executed in command mode and `:` might not guarantee same state of command mode I assume . But incase of lua why can't we directly call the lua callback in time of map evaluation ? What's the added benefit of K_COMMAND like structure in this case ?

zeertzjq commented:
I think some operations may be not safe to do when the mapping is evaluated.

bfredl commented:
@shadmansaleh K_COMMAND has nothing to with "command mode", on the contrary it executes a command safely according to the rules of each mappable mode. `<expr>` style execution has a lot of limitations like:
```
- Changing the buffer text |textlock|.
- Editing another buffer.
- The |:normal| command.
- Moving the cursor is allowed, but it is restored afterwards.
- If the cmdline is changed, the old text and cursor position are restored.
```
which it would be risky to try to disable as code calling into `getchar()` will not be prepared for them to happen.

shadmansaleh commented:
In that case I misunderstood it . Thanks @zeertzjq @bfredl

bfredl commented on src/nvim/api/vim.c:
```diff
@@ -1565,11 +1565,14 @@ ArrayOf(Dictionary) nvim_get_keymap(String mode)
 ///               Values are Booleans. Unknown key is an error.
 /// @param[out]   err   Error details, if any.
-void nvim_set_keymap(String mode, String lhs, String rhs, Dict(keymap) *opts, Error *err)
+void nvim_set_keymap(String mode, String lhs, Object rhs, Dict(keymap) *opts, Error *err)
```
Type of existing param cannot change in a stable API method, we'll need to use `opts` in this case.

shadmansaleh replied:
Isn't this still backward compatible ?

bfredl commented on src/nvim/getchar.c:
```diff
@@ -4560,3 +4626,47 @@ char_u *getcmdkeycmd(int promptc, void *cookie, int indent, bool do_concat)
   return (char_u *)line_ga.ga_data;
 }
+
+bool nlua_execute_keymap()
```
`nlua` is not the scope, so it should be `map_execute_lua` or something, also needs `void` to make it a prototype.

zeertzjq commented:
Oh, there are merge conflicts now.

bfredl commented on src/nvim/getchar.c:
```diff
@@ -2638,22 +2644,25 @@ void set_maparg_lhs_rhs(const char_u *orig_lhs, const size_t orig_lhs_len, const
-  mapargs->orig_rhs_len = orig_rhs_len;
-  mapargs->orig_rhs = xcalloc(mapargs->orig_rhs_len + 1, sizeof(char_u));
-  STRLCPY(mapargs->orig_rhs, orig_rhs, mapargs->orig_rhs_len + 1);
+  if (rhs_lua <= 0) {
```
use `LUA_NOREF` directly as sentinel value: `if (rhs_lua == NO_REF) ...`

bfredl commented on src/nvim/api/keysets.lua:
```diff
@@ -29,6 +29,9 @@ return {
     "script";
     "expr";
     "unique";
+    "callback";
+    "cb";
```
These two look duplicate, we should settle for one. (normally we would use `on_xx` for callback args though I am not sure what xx would be..)

shadmansaleh replied:
`cb` is just alias for callback . like set_hl can have fg/bg as alias.

I'm not sure if `on_xx` is appropriate here .

bfredl replied:
> like set_hl can have fg/bg as alias.

This is the special case. It doesn't mean we should go around and add lot's of aliases to freshly designed API:s. Let's pick one or the other. We are expected to create an "ergonomic" layer above this for init.lua anyway.

shadmansaleh replied:
ok then . will stick to callback only

gpanders commented on src/nvim/getchar.c:
```diff
@@ -3447,6 +3513,11 @@ static void showmap(mapblock_T *mp, bool local)
   }
+
+  if (mp->m_desc != NULL) {
+    msg_puts("\n                 ");  // Shift line to same level as rhs.
+    msg_puts(mp->m_desc);
```
Not sure about putting `desc` on a newline. If plugins or scripts depend on there being one mapping per line this would break that.

Knowing the number of the Lua ref is (imo) not that useful or interesting. What if instead we indicate a mapping uses a Lua function by prepending it with an `L` or `#` (similar to how buffer-local mappings have a `@` and non-remappable mappings use `*`). So e.g. `:map` would show

```
n  ZT          L Open a new tab
```

This should be pretty straightforward to add. The relevant code is at line 3486 in `getchar.c` and the docs under `|map-listing|`.

shadmansaleh replied:
But non lua keymaps can also have description . And lua keymap without description has to be shown as ref no any way . description being on separate line personally seems nicer.

Also I can't see why someone would be parsing `:map` command output when `maparg()` exists ?

gpanders replied:
>Also I can't see why someone would be parsing :map command output when maparg() & nvim_get_keymap exists ?

Just because it doesn't make sense doesn't mean people don't do it :) Though I'm not sure anyone actually is doing it, so maybe my concerns aren't valid.

But you're right, `desc` can be provided for non-Lua mappings as well, so this is probably fine.

shadmansaleh commented on src/nvim/api/private/helpers.c:
```diff
@@ -604,6 +605,12 @@ void modify_keymap(Buffer buffer, bool is_unmap, String mode, String lhs, String
   }
 
+  if (opts != NULL && opts->callback.type == kObjectTypeLuaRef) {
+    lua_funcref = opts->callback.data.luaref;
```
If new_ref is done here the ref can only be cleared only on fail as the ref will be attached to keymap on success . freeing it at end of this function will make keymap fail.

gpanders replied:
Yes, you only call `NLUA_CLEAR_REF` in the `fail_and_free` label, which is only used on error. If an error doesn't occur the ref is not freed.

shadmansaleh replied:
fail_and_free is always executed not on just fail . but we can workarround that so ref doesn't get cleared on success.

shadmansaleh commented on src/nvim/api/private/helpers.c:
```diff
@@ -1073,8 +1093,19 @@ ArrayOf(Dictionary) keymap_array(String mode, buf_T *buf)
+          for (size_t j = 0; j < d.size; j++) {
+            if (strequal("callback", d.items[j].key.data)) {
+              d.items[j].value.type = kObjectTypeLuaRef;
+              d.items[j].value.data.luaref = api_new_luaref((LuaRef)d.items[j].value.data.integer);
```
No they won't leak . They are attached to keymap and will be cleaned when keymap gets removed or overwritten .

gpanders replied:
Yes but you are making a new copy of the ref with `api_new_luaref`. The ref attached to the keymap will be freed when the keymap is deleted, but in this function you are only listing existing mappings. So when are those new copies freed?

Basically, does this work if you remove `api_new_luaref` and just copy the integer value directly?

shadmansaleh replied:
Ignore my earlier comment I thought this was on different part of the code.

The entire return object including all the luarefs of this function will be cleared at end of  `nvim_get_keymap` after after result is pushed to lua stack .

shadmansaleh commented on src/nvim/getchar.c:
```diff
@@ -15,6 +15,7 @@
 #include <stdbool.h>
 #include <string.h>
 
+#include "nvim/api/private/helpers.h"
```
I'm not sure I understamd what you're saying her . where should free_object & clear_error happen ?

shadmansaleh replied:
same as https://github.com/neovim/neovim/pull/16594#discussion_r776391779 ?

gpanders replied:
Yes, per bfredl's comment this is fine.

shadmansaleh commented on src/nvim/getchar.c:
```diff
@@ -4560,3 +4656,47 @@ char_u *getcmdkeycmd(int promptc, void *cookie, int indent, bool do_concat)
   return (char_u *)line_ga.ga_data;
 }
+
+bool map_execute_lua(void)
```
vgetorpeek cannot be called out of getchar . exposing it to  lua/executor feels worse.

rename from nlua_execute_map to map_execute_lua was suggested by bfredl

bfredl replied:
this uses global state of the mapping engine, so I think getchar.c is more appropriate as it is. The nlua part is already abstracted as `nlua_call_ref`

gpanders replied:
>vgetorpeek cannot be called out of getchar . exposing it to lua/executor feels worse.

Ok fair enough, but including `private/helpers.h` still doesn't seem right to me, judging by the "private" in the path.

bfredl replied:
It's an historical artifact that should be removed now that we can mix API and non-API functions in the same file. until then, including `private/helpers.h` to use the functions there is fine.

shadmansaleh committed:
feat(api): add support for lua function & description in keymap

Behavioral changes:

1. Added support for lua function in keymaps in
--------------------------------------------
- nvim_set_keymap
  Can set lua function as keymap rhs like following:
```lua
vim.api.nvim_{buf_}set_keymap('n', '<leader>lr', '', {callback = vim.lsp.buf.references})
```
  Note: lua function can only be set from lua . If api function being
  called from viml or over rpc this option isn't available.
- nvim_{buf_}get_keymap
  When called from lua, lua function is returned is `callback` key .
  But in other cases callback contains number of the function ref.
- :umap, nvim_del_keymap & nvim_buf_del_keymap clears lua keymaps correctly.
- :map commands for displaing rhs .
   For lua keymaps rhs is displayed as <Lua function ref_no>
   Note: lua keymap cannot be set through viml command / functions.
- mapargs()
  When dict is false it returns string in `<Lua function ref_no>`
  format (same format as :map commands).
  When dict is true it returns ref_no number in `callback` key.
- mapcheck()
  returns string in `<Lua function ref_no>` format (same format as :map commands).

2. Added support for keymap description
---------------------------------------
- nvim_{buf_}set_keymap: added `desc` option in opts table .
 ```lua
vim.api.nvim_set_keymap('n', '<leader>w', '<cmd>w<cr>', {desc='Save current file'})
```
- nvim_{buf_}get_keymap: contains `desc` in returned list.
- commands like `:nmap <leader>w` will show description in a new line below rhs.
- `maparg()` return dict contains `desc`.

bfredl commented on src/nvim/getchar.c:
```diff
@@ -2638,22 +2644,36 @@ void set_maparg_lhs_rhs(const char_u *orig_lhs, const size_t orig_lhs_len, const
-    mapargs->rhs = xcalloc(mapargs->rhs_len + 1, sizeof(char_u));
-    STRLCPY(mapargs->rhs, replaced, mapargs->rhs_len + 1);
+    size_t buflen = 1024;
+    char *tmp_buf = xmalloc(buflen);
```
use a static buffer, like `char tmp_buf[64]`.

bfredl commented on src/nvim/getchar.c:
```diff
@@ -3017,17 +3037,29 @@ int buf_do_map(int maptype, MapArguments *args, int mode, bool is_abbrev, buf_T
               mp->m_mode &= ~mode;  // remove mode bits
               if (mp->m_mode == 0 && !did_it) {  // reuse entry
-                xfree(mp->m_str);
+                if (mp->m_str != NULL) {
```
`XFREE_CLEAR` already handles NULL so this if and the next should be removed.

bfredl commented on src/nvim/getchar.c:
```diff
@@ -3017,17 +3037,29 @@ int buf_do_map(int maptype, MapArguments *args, int mode, bool is_abbrev, buf_T
                 mp->m_script_ctx.sc_lnum += sourcing_lnum;
+                mp->m_desc = NULL;
+                if (args->desc != NULL) {
+                  mp->m_desc = (char *)vim_strsave((char_u *)args->desc);
```
`xstrdup` avoids two casts.

shadmansaleh commented:
Is this good to be merged ?

shadmansaleh commented on src/nvim/getchar.c:
```diff
@@ -3934,15 +3978,33 @@ static char_u *eval_map_expr(char_u *str, int c)
+      }
+      api_free_object(ret);
+    } else {
+      nlua_call_ref(mp->m_luaref, NULL, args, false, &err);
```
Yep this can be removed . I had forgotten about that.

shadmansaleh replied:
done

bfredl commented on src/nvim/eval/funcs.c:
```diff
@@ -6027,10 +6028,15 @@ static void get_maparg(typval_T *argvars, typval_T *rettv, int exact)
+    } else if (rhs_lua != LUA_NOREF) {
+      size_t msglen = 100;
+      char *msg = (char *)xmalloc(msglen);
+      snprintf(msg, msglen, "<Lua function %d>", mp->m_luaref);
```
should we use desc here if available? Maybe `<Lua function: %s>` with the desc instead of opaque int.

shadmansaleh replied:
I have thought having same representation like `:map` command will be better . Also makes it consistent with when desc isn't given . the desc is provided when `dict` option is used on `maparg`

bfredl replied:
> I have thought having same representation like :map command will be better . 

I agree, and shouldn't `:map` use desc as well? also seems more helpful than showing opaque number to the user.

shadmansaleh replied:
`:map` shows desc in 2nd line . Because non lua keymaps can also have desc I feel like it'll be weird if for lua desc apeared in place of rhs and for non lua it apeared somewhere else.

Currently it gets shown like this.
```
:map xx
lhs   flags   rhs
                    desc
```

bfredl replied:
ok, let's keep it like it is for now. we can always develop it further until 0.7 anyway.

bfredl commented:
LGTM after comments, also a squash would be good :]

clason commented:
@shadmansaleh I'm getting a weird error with the following mapping (from https://github.com/L3MON4D3/LuaSnip, but I don't see an issue from the plugin itself):
```lua
map( 'i', '<tab>', '',
{callback= function()
  if require"luasnip".expand_or_jumpable() then
    vim.schedule(require"luasnip".expand_or_jump())
  else
    return "<tab>"
  end
end,
expr = true })
```
If I type a snippet and try to expand it with `<tab>`, I get
```
E5108: Error executing lua: ...site/pack/packer/start/luasnip/lua/luasnip/util/util.lua:191: Failed to save undo information
stack traceback:
        [C]: in function 'nvim_buf_set_text'
        ...site/pack/packer/start/luasnip/lua/luasnip/util/util.lua:191: in function 'put'
        ...ite/pack/packer/start/luasnip/lua/luasnip/nodes/node.lua:55: in function 'put_initial'
        .../pack/packer/start/luasnip/lua/luasnip/nodes/snippet.lua:559: in function 'put_initial'
        .../pack/packer/start/luasnip/lua/luasnip/nodes/snippet.lua:366: in function 'trigger_expand'
        ...nvim/site/pack/packer/start/luasnip/lua/luasnip/init.lua:167: in function 'snip_expand'
        ...nvim/site/pack/packer/start/luasnip/lua/luasnip/init.lua:240: in function 'expand'
        ...nvim/site/pack/packer/start/luasnip/lua/luasnip/init.lua:287: in function 'expand_or_jump'
        /Users/clason/.config/nvim/plugin/mappings.lua:17: in function </Users/clason/.config/nvim/plugin/mappings.lua:15>
```
(the "non-expanding" case works fine).

clason commented:
(For those only interested in getting the LuaSnip keymap to work, you can of course do
```lua
vim.keymap.set( 'i', '<tab>', function()
  if not require"luasnip".expand_or_jump() then
    vim.api.nvim_feedkeys(vim.api.nvim_replace_termcodes('<tab>', true, false, true), 'n', true)
  end
end)
```
instead; but I still would like to understand what's going on with the "naive" translation.)

shadmansaleh commented:
@clason I think you're getting that error because `expr` keymaps are executed under `textlock`.

clason commented:
Yep, that matches my observations and sounds like a reasonable explanation.

But why is the equivalent expr-mapping in Vimscript working then? Does `textlock` not apply there? If so, that's a bit of a regression in the new (or rather, Neovim) API?

shadmansaleh commented:
textlock applies to viml version too .

> ``lua
>vim.keymap.set( 'i', '<tab>', 
>  'luasnip#expand_or_jumpable() ? "<Plug>luasnip-expand-or-jump" : "<tab>"',
>  { expr = true })
>```
Viml version that you've given doesn't do text edit under expr map


equivalent lua will probably be
```lua
vim.keymap.set( 'i', '<tab>', function()
  if require"luasnip".expand_or_jumpable() then
    return "<Plug>luasnip-expand-or-jump"
  else
    return "<tab>"
  end,
  { expr = true })
```

zeertzjq commented:
> just inserts a literal <Plug>luasnip-expand-or-jump

This is because `noremap = false` is only automatically set when rhs begins with `<plug>`, and in this case it doesn't.

clason commented:
OK, that makes sense; then that's just a gotcha that is good to know. Thanks for explaining!
